docs/project/how-to-write-an-upgradable-application.md

```diff
@@ -0,0 +1,104 @@
+
+This article describes a list of useful tips on how to write your application to make the upgrade to the new version of the Shopsys Framework as smooth as possible.
+
+[TOC]
+
+## Constructor overrides
+
+From time to time you need to completely override functionality, so you don't call parent method at all.
+This is especially dangerous in constructors as constructors don't need to be the same in parent–children class.
+We leverage this to be able to introduce a new optional dependency of the class.
+
+If you don't call `parent:__construct()` in your code, and we add initialization of some property and that initialization will not be called in your code.
+It's really useful to add a comment for your future self that parent is not called intentionally and check for those comments while upgrading.
+
+For example when we make this change in the new version
+
+```diff
+namespace Shopsys\FrameworkBundle\Model\Product\Parameter;
+
+class ParameterData {
+
+    // ...
+
+    public function __construct()
+    {
+        $this->name = [];
+        $this->visible = false;
++       $this->promoted = false;
+    }
+```
+
+you may encounter problems with following code (notice the `promoted` property will not be set)
+
+```php
+namespace App\Model\Product\Parameter;
+
+use Shopsys\FrameworkBundle\Model\Product\Parameter\ParameterData as BaseParameterData;
+
+class ParameterData extends BaseParameterData {
+
+    // ...
+
+    public function __construct()
+    {
+        // parent::__construct() not called intentionally to avoid setting parameter visibility 
+        $this->name = [];
+    }
+```
+
+Adding a simple comment like one above help you identify those places and quickly check them.
+Also, you may be sure this is intentional, and not a bug. 
+
+## Reported bug
+
+Sometimes you may find a bug or request a feature in Shopsys Framework, 
+but until reported (thanks for that!) issue/request is processed, and a new version is released you need to fix it in your code.
+
+Then it's really useful to make a notice in your code, so you can quickly decide in the future whether some parts of your code can be safely deleted as it's already resolved.
+
+```php
+    /**
+     * @param array $matches
+     * @param string|null $locale
+     * @return string
+     */
+    protected function replace(array $matches, ?string $locale): string
+    {
+        // ...
+
+        // $productClassName is required to be set manually until https://github.com/shopsys/shopsys/issues/1693 is resolved
+        $productClassName = 'Shopsys\FrameworkBundle\Model\Product\Product';
+        $imageViews = $this->imageViewFacade->getForEntityIds($productClassName, $this->getIdsForProducts($products));
+        $productActionViews = $this->productActionViewFacade->getForProducts($products);
+
+        // ...
+```
+
+## Deprecations
+
+After upgrade (or before upgrade to another version) you may want to check for deprecations and resolve those reported from the Shopsys namespace.
```

Comment:
I am thinking that it would be nice to add more specific information - how to check the deprecations? AFAIK, one can find them in the Symfony debug toolbar, right? Or is there a command that we can run to list the deprecations?

Reply:
added hint